File: backend/game/clock.py
"""Server-side backgammon clock helpers. Mirrors frontend/src/lib/clock.ts.

Backgammon clocks use "simple delay": each turn gives the active player a few
free seconds (the delay) before their reserve time starts draining. A player's
reserve is only charged for the time they spend beyond the delay on a turn.
"""

from game import clock


NAMED_PRESET_SECONDS_PER_POINT = {
    'fast': 30,
    'normal': 60,
    'slow': 120,
}
TURN_DELAY_MS = 10_000


def parse_time_control(preset_id, target_points=1):
    """Return (base_ms, delay_ms) for a preset id like 'normal', else None.

    'none', missing, or malformed ids mean no time limit. Legacy 'M+S' ids
    from previously stored rooms still parse.
    """
    if not preset_id or preset_id == 'none':
        return None
    if preset_id in NAMED_PRESET_SECONDS_PER_POINT:
        try:
            target = max(1, int(target_points))
        except (TypeError, ValueError):
            target = 1
        return (
            NAMED_PRESET_SECONDS_PER_POINT[preset_id] * target * 1_000,
            TURN_DELAY_MS,
        )
    try:
        minutes, delay_sec = preset_id.split('+')
        return (int(minutes) * 60_000, int(delay_sec) * 1_000)
    except (ValueError, AttributeError):
        return None


def active_player(state):
    """The color whose clock should tick, or None when the clock is stopped.

    - waiting / game_over / opening_roll / opening_result -> stopped
    - rolling                             -> the current player pays while choosing roll/double
    - doubling_offered                    -> the responder decides, so they pay time
    - otherwise                           -> whoever's turn it is
    """
    if not state:
        return None
    phase = state.get('phase')
    if phase in ('waiting', 'game_over', 'opening_roll', 'opening_result'):
        return None
    if phase == 'doubling_offered' and state.get('doubleOfferedBy'):
        return 'black' if state['doubleOfferedBy'] == 'white' else 'white'
    return state.get('turn')


def apply_transition(clock, prev_active, new_active, elapsed_ms, delay_ms):
    """Charge the outgoing player only the time spent beyond their delay.

    Moving within the delay costs nothing; the new active player's clock was
    frozen while the old one played, so it is unchanged.
    """
    result = dict(clock)
    if prev_active and prev_active != new_active:
        charged = max(0, elapsed_ms - delay_ms)
        result[prev_active] = max(0, result.get(prev_active, 0) - charged)
    return result


def compute_clock(stored, incoming, now_ms, preset_id, target_points=1):
    """Server-owned clock computation (simple delay).

    Returns (clock, turn_started_at, active, timed_out, deadline_ms). When
    there is no time limit, returns (None, None, None, False, None). `stored`
    is the previously saved state; `incoming` is the state the client just
    sent. Client-provided clock values are never trusted.
    """
    tc = parse_time_control(preset_id, target_points)
    if tc is None:
        return None, None, None, False, None
    base_ms, delay_ms = tc

    clock = dict(stored.get('clock') or {'white': base_ms, 'black': base_ms})
    turn_started_at = stored.get('turnStartedAt')
    new_active = active_player(incoming)
    timed_out = False

    stored_active = active_player(stored)
    if stored_active and turn_started_at is not None:
        elapsed = max(0, now_ms - turn_started_at)
        charged = max(0, elapsed - delay_ms)
        if charged >= clock.get(stored_active, 0):
            clock[stored_active] = 0
            return clock, turn_started_at, stored_active, True, None

    if turn_started_at is None:
        # Seed the bank only at match start. An existing bank is deliberately
        # preserved while the clock is stopped between games.
        if not stored.get('clock'):
            clock = {'white': base_ms, 'black': base_ms}
        turn_started_at = now_ms if new_active else None
    else:
        if stored_active and new_active and stored_active != new_active:
            elapsed = max(0, now_ms - turn_started_at)
            clock = apply_transition(clock, stored_active, new_active, elapsed, delay_ms)
            turn_started_at = now_ms
        elif stored_active and not new_active:
            elapsed = max(0, now_ms - turn_started_at)
            clock = apply_transition(clock, stored_active, None, elapsed, delay_ms)
            turn_started_at = None
        elif not stored_active and new_active:
            turn_started_at = now_ms
        elif not new_active:
            turn_started_at = None

    deadline_ms = None
    if (
        new_active
        and turn_started_at is not None
        and clock.get(new_active, 0) > 0
    ):
        deadline_ms = (
            turn_started_at
            + delay_ms
            + clock.get(new_active, 0)
        )
    if new_active and clock.get(new_active, 0) <= 0:
        timed_out = True

    return clock, turn_started_at, new_active, timed_out, deadline_ms


def deadline_for(state, preset_id, target_points=1):
    """Epoch ms when the active player's reserve would hit zero, or None.

    With simple delay the reserve only drains after the per-turn delay, so the
    deadline is turnStartedAt + delay + reserve.
    """
    tc = parse_time_control(preset_id, target_points)
    if tc is None:
        return None
    _, delay_ms = tc
    clock = state.get('clock')
    active = active_player(state)
    turn_started_at = state.get('turnStartedAt')
    if not clock or not active or turn_started_at is None:
        return None
    remaining = clock.get(active, 0)
    if remaining <= 0:
        return None
    return turn_started_at + delay_ms + remaining

File: frontend/src/lib/clock.ts
import type { GameState, Color } from "./backgammon/engine";

export interface TimeControl {
  base: number; // starting reserve per player, ms
  delay: number; // free seconds per turn before the reserve drains, ms
}

export interface TimeControlPreset {
  id: string;
  label: string;
  secondsPerPoint: number;
  delay: number;
}

export const TIME_CONTROL_PRESETS: TimeControlPreset[] = [
  { id: "none", label: "No limit", secondsPerPoint: 0, delay: 0 },
  { id: "fast", label: "Fast", secondsPerPoint: 30, delay: 10_000 },
  { id: "normal", label: "Normal", secondsPerPoint: 60, delay: 10_000 },
  { id: "slow", label: "Slow", secondsPerPoint: 120, delay: 10_000 },
];

export function parseTimeControl(
  id: string | null | undefined,
  targetPoints = 1,
): TimeControl | null {
  if (!id || id === "none") return null;
  const preset = TIME_CONTROL_PRESETS.find((p) => p.id === id);
  if (preset) {
    const target = Number.isFinite(targetPoints)
      ? Math.max(1, Math.trunc(targetPoints))
      : 1;
    return {
      base: preset.secondsPerPoint * target * 1_000,
      delay: preset.delay,
    };
  }
  // Legacy "M+S" ids from previously stored rooms still parse.
  const [minutes, delaySec] = id.split("+").map((n) => parseInt(n, 10));
  if (!Number.isFinite(minutes) || !Number.isFinite(delaySec)) return null;
  return { base: minutes * 60_000, delay: delaySec * 1_000 };
}

export function activePlayerOf(state: GameState | null): Color | null {
  if (!state) return null;
  if (
    state.phase === "waiting" ||
    state.phase === "game_over" ||
    state.phase === "opening_roll" ||
    state.phase === "opening_result"
  ) {
    return null;
  }
  if (state.phase === "doubling_offered" && state.doubleOfferedBy) {
    return state.doubleOfferedBy === "white" ? "black" : "white";
  }
  return state.turn;
}

export function applyClockTransition(
  clock: Record<Color, number>,
  prevActive: Color | null,
  newActive: Color | null,
  elapsedMs: number,
  delayMs: number,
): Record<Color, number> {
  if (!prevActive || prevActive === newActive) return { ...clock };
  const charged = Math.max(0, elapsedMs - delayMs);
  return {
    ...clock,
    [prevActive]: Math.max(0, clock[prevActive] - charged),
  };
}

/** Reserve ms still available for the active player, accounting for the delay. */
export function reserveLeft(
  clock: Record<Color, number>,
  active: Color | null,
  turnStartedAt: number | null | undefined,
  nowMs: number,
  delayMs: number,
): number {
  if (!active) return 0;
  if (turnStartedAt == null) return clock[active] ?? 0;
  const charged = Math.max(0, nowMs - turnStartedAt - delayMs);
  return Math.max(0, (clock[active] ?? 0) - charged);
}

/** Delay ms remaining for the current turn (the "extra time for the play"). */
export function delayLeft(
  active: Color | null,
  turnStartedAt: number | null | undefined,
  nowMs: number,
  delayMs: number,
): number {
  if (!active || turnStartedAt == null) return 0;
  return Math.max(0, delayMs - (nowMs - turnStartedAt));
}

export function formatClock(ms: number | null | undefined): string {
  if (ms == null) return "--:--";
  const totalSec = Math.max(0, Math.ceil(ms / 1000));
  const m = Math.floor(totalSec / 60);
  const s = totalSec % 60;
  return `${m}:${String(s).padStart(2, "0")}`;
}
